CoreFoundation/PlugIn.subproj/CFBundle_Executable.c

```diff
@@ -303,38 +275,96 @@ static CFURLRef _CFBundleCopyExecutableURLInDirectory2(CFBundleRef bundle, CFURL
+    // - Freestanding bundles (`Bundle.resource` as sibling of the library/executable)
+    // - FHS bundles (`prefix/share/Bundle.resource`)
+    //
+    // Note:
+    // For freestanding and FHS bundles, this function has to support removing library/executable prefix/suffix
+    // on each platform. For example, it has to detect to the bundle name as `Bundle` for executables
+    // like `Bundle.exe` in Windows or `libBundle.so` on Linux.
     
-    buffLen = CFStringGetLength(str);
-    if (buffLen > CFMaxPathSize) buffLen = CFMaxPathSize;
-    CFStringGetCharacters(str, CFRangeMake(0, buffLen), buff);
+    CFURLRef url = NULL;
     
-#if DEPLOYMENT_TARGET_WINDOWS
-    // Is this a .dll or .exe?
-    if (buffLen >= 5 && (_wcsnicmp((wchar_t *)&(buff[buffLen-4]), L".dll", 4) == 0 || _wcsnicmp((wchar_t *)&(buff[buffLen-4]), L".exe", 4) == 0)) {
-        CFIndex extensionLength = CFStringGetLength(_CFBundleSiblingResourceDirectoryExtension);
-        buffLen -= 4;
-        // If this is an _debug, we should strip that before looking for the bundle
-        if (buffLen >= 7 && (_wcsnicmp((wchar_t *)&buff[buffLen-6], L"_debug", 6) == 0)) buffLen -= 6;
+#if FREESTANDING_BUNDLES
+    if (!url) {
+        CFURLRef executableURL = CFURLCreateWithFileSystemPath(kCFAllocatorSystemDefault, str, PLATFORM_PATH_STYLE, false);
+        CFStringRef executableName = CFURLCopyLastPathComponent(executableURL);
+        CFStringRef bundleName = NULL;
         
-        if (buffLen + 1 + extensionLength < CFMaxPathSize) {
-            buff[buffLen] = '.';
-            buffLen ++;
-            CFStringGetCharacters(_CFBundleSiblingResourceDirectoryExtension, CFRangeMake(0, extensionLength), buff + buffLen);
-            buffLen += extensionLength;
-            outstr = CFStringCreateWithCharactersNoCopy(kCFAllocatorSystemDefault, buff, buffLen, kCFAllocatorNull);
-            url = CFURLCreateWithFileSystemPath(kCFAllocatorSystemDefault, outstr, PLATFORM_PATH_STYLE, true);
-            CFRelease(outstr);
+        // Check if the executable has shared library prefix/suffix (libCF.dylib, libCF.so, CF.dll).
+        if (!bundleName) {
+            bundleName = _CFStringRemovePrefixSuffix(executableName,
+                                                     _CFBundleSharedLibraryFilenamePrefix,
+                                                     _CFBundleSharedLibraryFilenameSuffix);
+        }
+        
+        // Check if the executable has executable-style prefix/suffix (CF, CF.exe).
+        if (!bundleName) {
+            bundleName = _CFStringRemovePrefixSuffix(executableName,
+                                                     _CFBundleExecutableFilenamePrefix,
+                                                     _CFBundleExecutableFilenameSuffix);
+        }
+        
+        // Otherwise, use the executable name as bundle name (CF).
+        if (!bundleName) {
+            bundleName = executableName;
```

Comment:
Add a `CFRetain(executableName)` here and `CFRelease(bundleName)` below to avoid leaking.